sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/BootstrapperImpl.java

```diff
@@ -32,45 +35,53 @@ public BootstrapperImpl(PartitionSynchronizer synchronizer, LeaseStore leaseStor
         this.leaseStore = leaseStore;
         this.lockTime = lockTime;
         this.sleepTime = sleepTime;
+
+        this.isInitialized = false;
     }
 
     @Override
     public Mono<Void> initialize() {
         BootstrapperImpl self = this;
+        self.isInitialized = false;
 
-        return Mono.fromRunnable( () -> {
-            while (true) {
-                boolean initialized = self.leaseStore.isInitialized().block();
-
-                if (initialized) break;
+        return Mono.just(self)
+            .flatMap( value -> self.leaseStore.isInitialized())
+            .flatMap(initialized -> {
+                self.isInitialized = initialized;
 
-                boolean isLockAcquired = self.leaseStore.acquireInitializationLock(self.lockTime).block();
+                if (initialized) {
+                    return Mono.empty();
+                } else {
+                    return self.leaseStore.acquireInitializationLock(self.lockTime)
+                        .flatMap(lockAcquired -> {
+                            self.isLockAcquired = lockAcquired;
 
-                try {
-                    if (!isLockAcquired) {
-                        logger.info("Another instance is initializing the store");
-                        try {
-                            Thread.sleep(self.sleepTime.toMillis());
-                        } catch (InterruptedException ex) {
-                            logger.warn("Unexpected exception caught", ex);
-                        }
-                        continue;
-                    }
-
-                    logger.info("Initializing the store");
-                    self.synchronizer.createMissingLeases().block();
-                    self.leaseStore.markInitialized().block();
-
-                } catch (RuntimeException ex) {
-                    break;
-                } finally {
-                    if (isLockAcquired) {
-                        self.leaseStore.releaseInitializationLock().block();
-                    }
+                            if (!self.isLockAcquired) {
+                                logger.info("Another instance is initializing the store");
+                                try {
+                                    Thread.sleep(self.sleepTime.toMillis());
```

Comment:
this is blocking call!

Reply:
In this particular case we want to "block" the current thread since the lock was acquired by another instance of the CFP and we have to wait until it finishes initializing the lease container.

I'll call "delayElement()" on the return as an alternative to Thread.sleep().